### PR TITLE
Fix async-over-sync issue in Image.DecodeAsync()

### PR DESCRIPTION
**Pull request by kroymann**
### Prerequisites

- [X] I have written a descriptive pull-request title
- [X] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [X] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [X] I have provided test coverage for my change (where applicable)

### Description
While investigating recent instability in our production server environment, we discovered another code path within the async image loading code that would result in a blocking, synchronous stream read, which on our servers was leading to threadpool exhaustion.  This change adds a proper asynchronous variant of the offending code so that the stream reading triggered by `Image.DecodeAsync()` executes asynchronously.

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
Thanks @kroymann I'll have a proper read over the changes ASAP.

**Review by antonfirsov — COMMENTED** [review]
I think the following overload should be also updated to use  `InternalDetectFormatAsync`:

https://github.com/kroymann/ImageSharp/blob/9c5a393650bc40928caa89765501fdaf7bea1484/src/ImageSharp/Image.FromStream.cs#L68-L76

File: src/ImageSharp/Image.Decode.cs [hunk 1]
```diff
@@ -166,12 +233,12 @@ private static ImageInfo InternalIdentify(DecoderOptions options, Stream stream)
     /// <param name="stream">The stream.</param>
     /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
     /// <returns>The <see cref="ImageInfo"/>.</returns>
-    private static Task<ImageInfo> InternalIdentifyAsync(
+    private static async Task<ImageInfo> InternalIdentifyAsync(
```
[the remark this answers is not in the document]
Comment (kroymann) [reply src/ImageSharp/Image.Decode.cs:236]: To make this return a `ValueTask`, I think we would need/want to make `IdentifyAsync()` also return a `ValueTask`, and given that that API is public, I'm assuming we probably don't want to change its return type as that would be a binary breaking change, right?

[the remark this answers is not in the document]
Comment (kroymann) [reply src/ImageSharp/Image.Decode.cs:236]: I updated the overload you called out to use `InternalDetectFormatAsync` as suggested, but I left the return types (and `WithSeekableStreamAsync` alone as I didn't want to make a breaking change to any public API signatures.

**Comment by kroymann** [comment]
@antonfirsov Can you tell why the builds are failing after I added the two commits addressing your PR feedback?  They seem like innocuous changes, and I can't understand why the builds are now failing when they succeeded before.

**Comment by antonfirsov** [comment]
```
---- System.DllNotFoundException : Unable to load shared library 'libgdiplus' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: 
dlopen(/Users/runner/work/ImageSharp/ImageSharp/artifacts/bin/tests/ImageSharp.Tests/Release/net8.0/runtimes/osx-arm64/native/libgdiplus.dylib, 0x0001): tried: '/Users/runner/work/ImageSharp/ImageSharp/artifacts/bin/tests/ImageSharp.Tests/Release/net8.0/runtimes/osx-arm64/native/libgdiplus.dylib' (no such file), 
```

@kroymann this doesn't look related to your change. There is probably an issue with the `macos-latest` image, since restarting didn't help.

**Comment by antonfirsov** [comment]
#2728 should fix the mac runs.

File: src/ImageSharp/Image.Decode.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Image.Decode.cs:236]: I don't see the internal change, did you miss a commit? If not, don't bother, the impact is negligible.

**Review by antonfirsov — APPROVED** [review]
LGTM assuming all CI jobs gonna be green.

File: src/ImageSharp/Image.Decode.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Image.Decode.cs:236]: main is the V4 branch so we can break if need be.

[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Image.Decode.cs:236]: Thanks for the update; let’s get this merged then.

**Pull request merged**